File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlayQueueItemTouchCallback.java
package org.schabi.newpipe.player.playqueue;

import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

public abstract class PlayQueueItemTouchCallback extends ItemTouchHelper.SimpleCallback {
    private static final int MINIMUM_INITIAL_DRAG_VELOCITY = 10;
    private static final int MAXIMUM_INITIAL_DRAG_VELOCITY = 25;

    public PlayQueueItemTouchCallback() {
        super(ItemTouchHelper.UP | ItemTouchHelper.DOWN, ItemTouchHelper.RIGHT);
    }

    public abstract void onMove(int sourceIndex, int targetIndex);

    public abstract void onSwiped(int index);

    @Override
    public int interpolateOutOfBoundsScroll(final RecyclerView recyclerView, final int viewSize,
                                            final int viewSizeOutOfBounds, final int totalSize,
                                            final long msSinceStartScroll) {
        final int standardSpeed = super.interpolateOutOfBoundsScroll(recyclerView, viewSize,
                viewSizeOutOfBounds, totalSize, msSinceStartScroll);
        final int clampedAbsVelocity = Math.max(MINIMUM_INITIAL_DRAG_VELOCITY,
                Math.min(Math.abs(standardSpeed), MAXIMUM_INITIAL_DRAG_VELOCITY));
        return clampedAbsVelocity * (int) Math.signum(viewSizeOutOfBounds);
    }

    @Override
    public boolean onMove(final RecyclerView recyclerView, final RecyclerView.ViewHolder source,
                          final RecyclerView.ViewHolder target) {
        if (source.getItemViewType() != target.getItemViewType()) {
            return false;
        }

        final int sourceIndex = source.getLayoutPosition();
        final int targetIndex = target.getLayoutPosition();
        onMove(sourceIndex, targetIndex);
        return true;
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return false;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return true;
    }

    @Override
    public void onSwiped(final RecyclerView.ViewHolder viewHolder, final int swipeDir) {
        onSwiped(viewHolder.getAdapterPosition());
    }
}

File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlayQueueItemTouchCallback.kt
package org.schabi.newpipe.player.playqueue

import androidx.core.math.MathUtils
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import kotlin.math.abs
import kotlin.math.sign

abstract class PlayQueueItemTouchCallback() : ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN, ItemTouchHelper.RIGHT) {
    abstract fun onMove(sourceIndex: Int, targetIndex: Int)
    abstract fun onSwiped(index: Int)
    public override fun interpolateOutOfBoundsScroll(recyclerView: RecyclerView,
                                                     viewSize: Int,
                                                     viewSizeOutOfBounds: Int,
                                                     totalSize: Int,
                                                     msSinceStartScroll: Long): Int {
        val standardSpeed: Int = super.interpolateOutOfBoundsScroll(recyclerView, viewSize,
                viewSizeOutOfBounds, totalSize, msSinceStartScroll)
        val clampedAbsVelocity: Int = MathUtils.clamp(abs(standardSpeed.toDouble()),
                MINIMUM_INITIAL_DRAG_VELOCITY, MAXIMUM_INITIAL_DRAG_VELOCITY)
        return clampedAbsVelocity * sign(viewSizeOutOfBounds.toDouble()).toInt()
    }

    public override fun onMove(recyclerView: RecyclerView,
                               source: RecyclerView.ViewHolder,
                               target: RecyclerView.ViewHolder): Boolean {
        if (source.getItemViewType() != target.getItemViewType()) {
            return false
        }
        val sourceIndex: Int = source.getLayoutPosition()
        val targetIndex: Int = target.getLayoutPosition()
        onMove(sourceIndex, targetIndex)
        return true
    }

    public override fun isLongPressDragEnabled(): Boolean {
        return false
    }

    public override fun isItemViewSwipeEnabled(): Boolean {
        return true
    }

    public override fun onSwiped(viewHolder: RecyclerView.ViewHolder, swipeDir: Int) {
        onSwiped(viewHolder.getBindingAdapterPosition())
    }

    companion object {
        private val MINIMUM_INITIAL_DRAG_VELOCITY: Int = 10
        private val MAXIMUM_INITIAL_DRAG_VELOCITY: Int = 25
    }
}
